### PR TITLE
Amesos2: Refactor Klu2 solver to use Kokkos views.

### PR DESCRIPTION
**Pull request by MicheldeMessieres**
@trilinos/amesos2

## Motivation

Switch Klu2 solver to use views and test with device and host inputs.

## Testing

Amesos2_KLU2_Solver_Test_MPI_2 (new)
Amesos2_KLU2_UnitTests_MPI_2
Amesos2_SolverFactory_UnitTests_MPI_4

### CODE REVIEW

File: packages/amesos2/src/Amesos2_KLU2_def.hpp [hunk 1]
```diff
@@ -552,9 +519,12 @@ KLU2<Matrix,Vector>::loadA_impl(EPhase current_phase)
...
   // Only the root image needs storage allocated
   if( this->root_ ){
-    nzvals_.resize(this->globalNumNonZeros_);
-    rowind_.resize(this->globalNumNonZeros_);
-    colptr_.resize(this->globalNumCols_ + 1);
+    host_nzvals_view_ = host_value_type_array(
+      Kokkos::ViewAllocateWithoutInitializing("nzvals"), this->globalNumNonZeros_);
+    host_rows_view_ = host_ordinal_type_array(
+      Kokkos::ViewAllocateWithoutInitializing("rowptr"), this->globalNumNonZeros_);
+    host_col_ptr_view_ = host_ordinal_type_array(
```
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_KLU2_def.hpp:526]: Minor comment, looks like View label convention is inconsistent with the variable name (e.g. `host_col_ptr_view_` with label `colind`, similar above for rows and `rowptrs`

File: packages/amesos2/src/Amesos2_KLU2_def.hpp [hunk 2]
```diff
@@ -304,44 +236,70 @@ KLU2<Matrix,Vector>::solve_impl(
...
+            ROOTED, this->rowIndexBase_);
+    }
+    else {
+      Util::get_1d_copy_helper_kokkos_view<MultiVecAdapter<Vector>,
+        host_solve_array_t>::do_get(B, bValues_,
+            as<size_t>(ld_rhs),
+            CONTIGUOUS_AND_ROOTED, this->rowIndexBase_);
+    }
+
+    // see Amesos2_Tacho_def.hpp for an explanation of why we 'get' X
```
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_KLU2_def.hpp:257]: Just checking that this is intended to refer back to the Tacho code, and not copy/paste

File: packages/amesos2/src/Amesos2_KLU2_def.hpp (hunk 1 above)
Comment (MicheldeMessieres) [reply packages/amesos2/src/Amesos2_KLU2_def.hpp:526]: Thanks. I will fix in the upcoming Basker PR unless auto test fails here.

File: packages/amesos2/src/Amesos2_KLU2_def.hpp (hunk 2 above)
Comment (MicheldeMessieres) [reply packages/amesos2/src/Amesos2_KLU2_def.hpp:257]: Correct. I didn't want to duplicate large comments on this. I have an issue open #7158 to address these and will do all the solvers together. I have some ideas how to handle this and then will clean up all the comments together.

**Pull request merged**